### DISCUSSION TITLE
GridGeoSampler example

### DISCUSSION THREAD

**Discussion by roybenhayun** [torchgeo/torchgeo#1629] [Q&A]
Hi, this might be related to [a previous post](https://github.com/microsoft/torchgeo/discussions/1624)...

due to constraints in the number and sizes of labeled data, we are trying to use GridGeoSampler instead of RandomBatchGeoSampler. 

tried to use patch size 32 and replace RandomBatchGeoSampler in the following SegmentationDataModule code:
```
    def train_dataloader(self):
        grid_sampler = GridGeoSampler(self.ds, size=self.patch_size, stride=self.patch_size-5)
        return DataLoader(self.ds, sampler=grid_sampler, collate_fn=stack_samples)
        # replaced:
        # sampler = RandomBatchGeoSampler(self.ds, size=self.patch_size, batch_size=self.batch_size,
        #                                 length=self.train_batches_per_epoch * self.batch_size)
        #
        # return DataLoader(self.ds, batch_sampler=sampler, num_workers=self.num_workers, collate_fn=stack_samples)

    def val_dataloader(self):
        grid_sampler = GridGeoSampler(self.ds, size=self.patch_size, stride=self.patch_size-5)
        return DataLoader(self.ds, sampler=grid_sampler, collate_fn=stack_samples)
        # replaced:

```
things start to work e.g., sanity checks pass incl. the datamodule plot() being called with samples. but then at the first batch getting:
`ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 1024, 1, 1])`

would be happy for a clarification or suggestion.
an example of GridGeoSampler, not as a standalone example which we already have, but in a training example, would be great.

thanks!

**Comment by roybenhayun** [comment]
FWIW, happens in 
File "C:\ProgramData\anaconda3\envs\ce_ai\lib\site-packages\torch\nn\functional.py", line 2416, in _verify_batch_size

**Comment by adamjstewart** [comment]
> due to constraints in the number and sizes of labeled data, we are trying to use GridGeoSampler instead of RandomBatchGeoSampler.

GridGeoSampler will not give you more samples. In fact, it will give you fewer samples. You should use RandomBatchGeoSampler during training and GridGeoSampler during evaluation.

> tried to use patch size 32 and replace RandomBatchGeoSampler in the following SegmentationDataModule code

I don't know what `SegmentationDataModule` is, can you give me the full code and possibly data so I can copy-n-paste and reproduce the issue?

> an example of GridGeoSampler, not as a standalone example which we already have, but in a training example, would be great.

We have examples in several builtin DataModules, including Chesapeake, L7 Irish, L8 Biome, and NAIP. But I agree that a tutorial example would be nice. We're planning on expanding our tutorials in the near future.

**Reply by roybenhayun** [reply]
yes, it will give less examples. sorry, forgot to mention that with patch sizes 256, 128, 64 it worked. with much less examples as you noted ofcourse. however, some contraints etc, we really want to try also 32.

the segmentation data module is based on [training-example](https://github.com/microsoft/satellite-imagery-labeling-tool/blob/main/example/training-example.ipynb). almost identical. 

just thinking... we hope to use augmentation to enrich. still learning how to use it e.g., rotating, scaling etc. so if patchs size 64 works, and we could tweak some of the masks we have, and augmentation would help to give more samples... will that be a good workaround? an example for augmentation would be great also.

**Reply by adamjstewart** [reply]
> the segmentation data module is based on training-example. almost identical.

That notebook is almost a year old and 2 releases of TorchGeo behind. It's also not part of this repo. But maybe you can open an issue on that repo.

**Comment by roybenhayun** [comment]
> I don't know what `SegmentationDataModule` is, can you give me the full code and possibly data so I can copy-n-paste and reproduce the issue?

that's much appreciated. will ping you on the slack channel. thanks.
